src/library/coap.cpp

```diff
@@ -995,11 +993,30 @@ std::shared_ptr<Message> Message::Deserialize(Error &aError, const ByteArray &aB
 
         SuccessOrExit(error = Deserialize(number, value, lastOptionNumber, aBuf, offset));
 
-        // Stop if any unrecognized option is critical.
-        // Otherwise, we need to proceed to decoding following options.
-        VerifyOrExit(IsValidOption(number, value) || !IsCriticalOption(number), error = Error::kBadOption);
+        if (IsValidOption(number, value))
+        {
+            // AppendOption will do further validation before adding the option.
+            error = message->AppendOption(number, value);
+        }
+        else
+        {
+            error = Error::kBadFormat;
+        }
+
+        if (error != Error::kNone)
+        {
+            if (IsCriticalOption(number))
+            {
+                // Stop if any unrecognized option is critical.
+                ExitNow(error = Error::kBadOption);
+            }
+            else
+            {
+                // Ignore non-critical option error.
+                error = Error::kNone;
+            }
+        }
```

Comment:
If `IsValidOption(number, value) == false` and `IsCriticalOption(number) == false`, `message->AppendOption(number, value)` would be executed before, but not after this change.

Is this change of behavior intended?

Reply:
@Moandor-y thanks for you careful consideration. Actually, `AppendOption` itself does arguments validation and rejects if `IsValidOption(number, value) == false`. Previous implementation ignores the error returned by `AppendOption` silently. This PR should not change the actual behavior but explicitly ignore the error by setting it to `Error::kNone`.